### PR TITLE
CMakePackage pass python hints automatically

### PR DESCRIPTION
**Pull request by kwryankrattiger**
Pass python hints automatically to cmake packages to help `FindPython` find new versions of python. Can be disabled by setting the `find_python_hints=False` in the package.

CC: @haampie @johnwparent

### CODE REVIEW

**Comment by wdconinc** [comment]
In a case of cosmic irony, I encountered this issue today after commenting on the slack thread. With `develop` it finds a system python:
```
     62      Could NOT find Python: Found unsuitable version "3.10", but required is
     63      exact version "3.10.13" (found /usr/include/python3.11, )
```
And that issue is fixed with this PR.

File: lib/spack/spack/build_systems/cmake.py [hunk 1]
```diff
@@ -288,6 +291,15 @@ def std_args(pkg, generator=None):
                 [define("CMAKE_FIND_FRAMEWORK", "LAST"), define("CMAKE_FIND_APPBUNDLE", "LAST")]
             )
 
+        if getattr(pkg, "find_python_hints", False) and pkg.spec.satisfies("^python"):
```
Comment (haampie) [inline lib/spack/spack/build_systems/cmake.py]: as i said on slack look at direct deps

File: lib/spack/spack/build_systems/cmake.py [hunk 2]
```diff
@@ -288,6 +291,15 @@ def std_args(pkg, generator=None):
                 [define("CMAKE_FIND_FRAMEWORK", "LAST"), define("CMAKE_FIND_APPBUNDLE", "LAST")]
             )
 
+        if getattr(pkg, "find_python_hints", False) and "python" in pkg.dependency_names():
```
Comment (haampie) [inline lib/spack/spack/build_systems/cmake.py]: This is wrong. Don't use `pkg`, use `spec`. E.g. `if spec.dependencies("python")`.

**Review by haampie — DISMISSED** [review]

**Comment by kwryankrattiger** [comment]
Changes by @haampie look fine, I wasn't sure what you meant by dependencies, makes sense.

+1

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 3]
```diff
@@ -540,12 +540,10 @@ def cmake_args(self):
...
             options.append(self.define_from_variant("DEAL_II_COMPONENT_PYTHON_BINDINGS", "python"))
             if "+python" in spec:
-                python_exe = spec["python"].command.path
                 python_library = spec["python"].libs[0]
                 python_include = spec["python"].headers.directories[0]
                 options.extend(
                     [
-                        self.define("PYTHON_EXECUTABLE", python_exe),
                         self.define("PYTHON_INCLUDE_DIR", python_include),
                         self.define("PYTHON_LIBRARY", python_library),
```
Comment (wdconinc) [inline var/spack/repos/builtin/packages/dealii/package.py]: Are these still needed? See `clingo` above where you removed them.

Comment (kwryankrattiger) [reply var/spack/repos/builtin/packages/dealii/package.py]: Maybe not. I am currently cleaning up the last patch and will come back around to these and test them. There are a few failures in CI and a few more failures that didn't come up in CI but are issues that need to be resolved.

Comment (haampie) [reply var/spack/repos/builtin/packages/dealii/package.py]: Yeah, these extra hints are probably unnecessary. Others that determine the install dir could probably stay.

Comment (haampie) [reply var/spack/repos/builtin/packages/dealii/package.py]: Actually it depends. If `find_package(PythonLibs)` is called before `find_package(PythonInterp)` we need to set those.

**Comment by haampie** [comment]
Failure seems unrelated.

> ==> Error: Failed to install openpmd-api due to InstallError: No binary found when cache-only was specified

File: lib/spack/spack/build_systems/cmake.py [hunk 4]
```diff
@@ -31,8 +32,30 @@ def _extract_primary_generator(generator):
...
+    """Set the PYTHON_EXECUTABLE, Python_EXECUTABLE, and Python3_EXECUTABLE CMake variables
+    if the package has Python as build or link dep and ``find_python_hints`` is set to True. See
+    ``find_python_hints`` for context."""
+    if not getattr(pkg, "find_python_hints", False):
+        return
+    pythons = pkg.spec.dependencies("python", dt.BUILD | dt.LINK)
+    if len(pythons) != 1:
+        return
+    try:
+        python_executable = pythons[0].package.command.path
```
Comment (kwryankrattiger) [inline lib/spack/spack/build_systems/cmake.py:48]: No need to fix here, but I would really like to know why python is showing up in the dependencies list...

**Comment by kwryankrattiger** [comment]
+1 to @haampie change

CI looks okay, failed push to mirror in upstream job is cause of CI failure

failed to push: https://gitlab.spack.io/spack/spack/-/jobs/9839540
no binary to pull: https://gitlab.spack.io/spack/spack/-/jobs/9839621

These types of issues are fixed in https://github.com/spack/spack/pull/40045

This should be safe to force merge.

Note: We should probably better understand why `python` is showing up in the `Spec.dependencies` of packages that do not have any dependence on `python`. But that shouldn't be a blocker for solving the existing issue with python detection in CMake packages.

**Comment by haampie** [comment]
> Note: We should probably better understand why `python` is showing up in the `Spec.dependencies` of packages that do not have any dependence on `python`

Forgot to mention that, but that is solved in #42215. We were setting build related globals on build deps that should (and *their* build deps are not installed, causing the issue)

**Pull request merged**